Prepare design doc ahead of balloting (#506)
unicode-org/message-format-wg: exploration/delimiting-variant-patterns.md

	</dl>
</details>

## Objective

The current syntax requires all patterns in _complex messages_ to be delimited ("quoted")
using curly brackets.
The working group has discussed at length different options for allowing patterns
to be unquoted.
This document contains the proposed solutions, including some that have been rejected by
the working group, along with some of the arguments related to the relative
suitability of each solution.

### Balloting Instructions

In the 2023-10-30 WG teleconference, there was a unanimous support for balloting the group on
the following question:

> Using [instant runoff voting](https://en.wikipedia.org/wiki/Instant-runoff_voting), rank
> your choice for how to handle delimiting of patterns in _complex messages_.

The **_deadline_** is **1700 (5 PM) in the `America/Los_Angeles` time zone on Saturday, 4 November, 2023**

- A group member in good standing MAY submit one vote consisting of a ranked set of choices
  up until the deadline.
- A group member MAY edit, change, or delete their vote up until the deadline.
- Votes MUST be submitted as a comment on github issue [#505](https://github.com/unicode-org/message-format-wg/issues/505).

>   Group members who cannot submit a comment on this issue should contact the chair (@aphillips) for assitance.

Reviewer: ```suggestion
  Group members who cannot submit a comment on this issue should contact the chair (@aphillips) for assistance.
```

- Votes MUST contain a stack ranked list of candidate options.
- Votes MUST only contain votes for candidate options.
  Write in votes are not acceptable.
- A vote MUST have at least one
  item in order to be counted and MAY rank two items or all three items.
- Only ranked votes will be counted. That is, do not submit a vote equating two entries.
- Group members MUST NOT comment on the votes of others in the voting thread.
  "Electioneering" or non-voting commentary is not permitted in the issue
  except for the chair seeking clarification of a vote.

### Definitions

**_complex messages_** are messages that contain _declarations_, _selectors_, or both according
to the ABNF.

**_simple messages_** are messages consisting only of a _pattern_, with no _declarations_ or _selectors_.

**_group member in good standing_** is any member of the MessageFormat mailing list or

> watcher of the message-format-wg github repo who has not be banned.

Reviewer: ```suggestion
watcher of the message-format-wg github repo who has not been banned.
```

- The quoting reduces the number of in-pattern escapes to the open/close sequence.
  and the placeholder sequence sigils.
- Since the pattern is already quoted, translators never have to add pattern quotes

>   in order to include whitespace into a given pattern.

Reviewer: ```suggestion
  in order to add leading or trailing whitespace in a given pattern.
```

Author: Mostly good, but "include"/"into" I think is slightly clearer.

```suggestion
in order to include leading or trailing whitespace into a given pattern.
```

- Has two ways to represent a pattern.
- May be difficult for translators to add quotes when needed.

### 4. Trim All Unquoted

In this alternative, non-simple patterns are trimmed
and we do not allow/provide for pattern quoting.

> Instead, pattern-sigificant whitespace must be individually quoted.

Reviewer: ```suggestion
Instead, pattern-significant whitespace must be individually quoted.
```

> Note that only the first character of whitespace has to be quoted
> or that an empty literal can be used to start or end the pattern.

Reviewer: ```suggestion
Note that only the first character of whitespace has to be quoted,
or an empty literal `{||}` can be used to start or end the pattern.
```

>{when many}   {| |}This also has one space start and end.{| |}
>{when *}
>
>       None of the whitespace before the letter 'N' at the start of this line
> is part of the pattern but all of the whitespace after the period at
> the end of this sentence is included in the pattern.
>
>       {||}
>```

Pros:
- Code is special, whitespace is not.

> - Makes whitespace into a "special event", alerting developers to the non-I18N aspects of it?

Reviewer: ```suggestion
- Makes leading and trailing whitespace into a "special event",
  alerting developers to the non-I18N aspects of it?
```
